### PR TITLE
Update phylo workflow

### PR DESCRIPTION
**Pull request by genehack**
## Description of proposed changes

Update the `phylogenetic` workflow to work with Nextclade v3.

Note that this is still a work in progress and the workflow still errors out; pushing my initial changes for an initial review and to get advice on next steps. 

## Related issue(s)

#2

### CODE REVIEW

**Comment by genehack** [comment]
Currently _some_ (but I think not all) of the `translate` jobs are failing with the following stacktrace: 

```
        augur translate             --tree results/hku1/tree.nwk             --ancestral-sequences results/hku1/nt_mu16:56:57 [55/29803]
 --reference-sequence config/hku1/genemap.gff             --output results/hku1/aa_muts.json
Traceback (most recent call last):
  File "/opt/miniconda3/envs/nextstrain-dev/lib/python3.8/site-packages/augur/__init__.py", line 66, in run
    return args.__command__.run(args)
  File "/opt/miniconda3/envs/nextstrain-dev/lib/python3.8/site-packages/augur/translate.py", line 418, in run
    features = load_features(args.reference_sequence, genes)
  File "/opt/miniconda3/envs/nextstrain-dev/lib/python3.8/site-packages/augur/utils.py", line 197, in load_features
    return _read_gff(reference, feature_names)
  File "/opt/miniconda3/envs/nextstrain-dev/lib/python3.8/site-packages/augur/utils.py", line 299, in _read_gff
    gff_entries = list(GFF.parse(in_handle, limit_info={'gff_type': valid_types}))
  File "/opt/miniconda3/envs/nextstrain-dev/lib/python3.8/site-packages/BCBio/GFF/GFFParser.py", line 791, in parse
    for rec in parser.parse_in_parts(gff_files, base_dict, limit_info,
  File "/opt/miniconda3/envs/nextstrain-dev/lib/python3.8/site-packages/BCBio/GFF/GFFParser.py", line 334, in parse_in_parts
    cur_dict = self._results_to_features(cur_dict, results)
  File "/opt/miniconda3/envs/nextstrain-dev/lib/python3.8/site-packages/BCBio/GFF/GFFParser.py", line 372, in _results_to_features
    (_, base) = self._add_toplevel_feature(base, feature)
  File "/opt/miniconda3/envs/nextstrain-dev/lib/python3.8/site-packages/BCBio/GFF/GFFParser.py", line 579, in _add_toplevel_feature
    new_feature = self._get_feature(feature_dict)
  File "/opt/miniconda3/envs/nextstrain-dev/lib/python3.8/site-packages/BCBio/GFF/GFFParser.py", line 589, in _get_feature
    new_feature = SeqFeature.SeqFeature(SeqFeature.SimpleLocation(start=rstart, end=rend, strand=feature_dict['strand']), feature_dict['
type'],
AttributeError: module 'Bio.SeqFeature' has no attribute 'SimpleLocation'


An error occurred (see above) that has not been properly handled by Augur.
To report this, please open a new issue including the original command and the error above:
    <https://github.com/nextstrain/augur/issues/new/choose>

[Tue May 14 16:56:56 2024]
Error in rule translate:
    jobid: 28
    input: results/hku1/tree.nwk, results/hku1/nt_muts.json, config/hku1/genemap.gff
    output: results/hku1/aa_muts.json
    shell:

        augur translate             --tree results/hku1/tree.nwk             --ancestral-sequences results/hku1/nt_muts.json
 --reference-sequence config/hku1/genemap.gff             --output results/hku1/aa_muts.json
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)
```

I'm guessing that it is probably not a coincidence that this is the GFF file that I modified...

File: phylogenetic/config/hku1/genemap.gff
Comment (victorlin) [inline phylogenetic/config/hku1/genemap.gff:1]: re: https://github.com/nextstrain/seasonal-cov/pull/3#issuecomment-2111351501

This error is https://github.com/nextstrain/augur/issues/1401. What version of `bcbio-gff` do you have installed?

**Comment by genehack** [comment]
> I'm guessing that it is probably not a coincidence that this is the GFF file that I modified...

Ope, perhaps I guessed wrong — reverting the `s/gene=/gene_name=/g` change I made to `hku1/genemap.gff` and manually running the `augur` command fails with the same error.

File: phylogenetic/config/hku1/genemap.gff
Comment (genehack) [reply phylogenetic/config/hku1/genemap.gff:1]:
```
> conda list | grep bcbio-gff
bcbio-gff                 0.7.1              pyh7cba7a3_1    bioconda
```

Comment (genehack) [reply phylogenetic/config/hku1/genemap.gff:1]: also, 
```
> conda list | grep -i biopython 
biopython                 1.74             py38h0b31af3_0    conda-forge
```

Comment (genehack) [reply phylogenetic/config/hku1/genemap.gff:1]: upgrading to this fixed the problem with `augur`: 

```
> conda list | grep -i biopython 
# packages in environment at /opt/miniconda3/envs/nextstrain-dev-test-biopython-upgrade:
biopython                 1.81             py38hae2e43d_1    conda-forge
```

Comment (genehack) [reply phylogenetic/config/hku1/genemap.gff:1]: ...and upgrading Biopython to that version allowed the workflow to complete. 

The specific command I ran was `conda install conda-forge::biopython`

Now the question is, how to capture this dependency information... ([relevant Slack discussion](https://bedfordlab.slack.com/archives/C0K3GS3J8/p1715732082600809))

**Comment by genehack** [comment]
These changes plus the forced BioPython version upgrade discussed above were sufficient to allow the workflow to complete, so I'm going to take this out of draft. I think the dependency management question is orthogonal enough to the original ticket that these changes (once reviewed and I'm sure modified) could get merged.

File: phylogenetic/config/hku1/genemap.gff
Comment (tsibley) [reply phylogenetic/config/hku1/genemap.gff:1]: Do you encounter this error with any of our managed runtimes? Or is is just in your ambient setup?

Comment (tsibley) [reply phylogenetic/config/hku1/genemap.gff:1]: I think that these two things being true at the same time

```
> conda list | grep bcbio-gff
bcbio-gff                 0.7.1              pyh7cba7a3_1    bioconda

> conda list | grep -i biopython 
biopython                 1.74             py38h0b31af3_0    conda-forge
```

means whatever solved the deps for your ambient environment did it wrong.

Maybe it didn't do it right initially, or maybe it messed it up with a subsequent install later that didn't account for existing deps in the environment, but [Augur pretty clearly declares](https://github.com/nextstrain/augur/blob/5d04da22723a308069b8b0ea457beb3f212b5e67/setup.py#L54-L56):

```
"bcbio-gff >=0.7.1, ==0.7.*",
"biopython >=1.80, ==1.*",
```

and that's also expressed in [the Conda packaging](https://github.com/bioconda/bioconda-recipes/blob/4a799838b65e7d7237f5db199fe930c5513eb2cb/recipes/augur/meta.yaml#L26-L27):

```
- bcbio-gff >=0.7.1,<0.8
- biopython >=1.80,<2
```

So this really doesn't look like a case of "pathogen repo needs to declare a specific dep" to me.

Comment (genehack) [reply phylogenetic/config/hku1/genemap.gff:1]: The dependency resolver is `miniconda`, specifically `conda 24.3.0`. On a different computer, where I set up a `nextstrain` environment in the same way, I _do_ have the correct version of `biopython`, so clearly on this laptop something got changed. 

One difference is that on this laptop (where the old `biopython` got installed), I had `conda install`d `snakefmt` into the environment, which _might_ be the proximate cause of this? There are a number of other gratuitous version differences (e.g., `augur 24.0.0` versus `augur 24.3.0`; `python 3.8.15` versus `python 3.10.13`) but `snakefmt` is the one glaring _addition_.

Comment (victorlin) [reply phylogenetic/config/hku1/genemap.gff:1]: Created an issue over in Augur to track cases like this, which aren't uncommon: https://github.com/nextstrain/augur/issues/1463

**Comment by huddlej** [comment]
@genehack I will agree with your comment in [4eeffd8](https://github.com/nextstrain/seasonal-cov/pull/3/commits/4eeffd8d9ef0b2d0205c33a12a32ddf82e5d20a9) that the format of the gene map file is a little confusing. In the official Nextclade datasets, I see [older style formatting with `gene` features and `gene_name` ids](https://github.com/nextstrain/nextclade_data/blob/release/data/nextstrain/sars-cov-2/BA.2.86/genome_annotation.gff3) and [the newer style of mostly `CDS` features with `gene_name` ids](https://github.com/nextstrain/nextclade_data/blob/release/data/nextstrain/rsv/a/EPI_ISL_412866/genome_annotation.gff3) and then [a version with `CDS` features and `gene` ids](https://github.com/nextstrain/nextclade_data/blob/release/data/nextstrain/mpox/all-clades/genome_annotation.gff3).

[The official Nextclade gene map docs](https://docs.nextstrain.org/projects/nextclade/en/stable/user/input-files/03-genome-annotation.html) suggest we should use `CDS` features with `gene` ids.

What you've done here looks reasonable, since most of our older workflows use `gene` features and `gene_name` ids, but maybe @ivan-aksamentov, @corneliusroemer, or @rneher have suggestions.

**Comment by ivan-aksamentov** [comment]
@genehack @huddlej 

I made a detailed writeup. This might go to the docs somewhere, though it is only useful for upgrading older datasets by the team internally, and is not super useful for most users. I think, for most users, saying "follow GFF3 spec" is more concise and less confusing.

Here is the full story:


### The GFF3 spec

The official GFF3 spec is here: https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md

The spec is complex, has many redundancies and degrees of freedom. A particular annotation can be achieved in many different ways. In that sense, it's less of a "format", as it does not specify the exact form or shape, but more of a "guide", providing a set of best practices, recommendations and constraints for a popular way of expressing annotations in the bioinformatics community, and aiming to unify the conflicting ad-hoc formats and forks which appeared over the years.

The TSV-like nature of the format should not mislead you - while some columns do define some basics, the most of the actually useful information is stuffed into the "attributes" column (column 9), which is a container for key-value pairs, some constrained, some arbitrary. Crazy things can (and do) happen there in the wild. And column 9 is the source of most breakages and of the complexity in the parsers.

In Nextclade v3, the closer you follow the GFF3 spec the higher the chances that things will work correctly. Please report any deviations as bugs by opening issues in Nextclade repo.


### Historical context

In older versions of Nextclade we used non-spec-compliant `gene` and `gene_name` keys in the `attributes` column (column 9). More over, older versions of Nextclade supported only flat genome annotations, meaning the only supported feature type was "gene", and the list was flat, i.e. there were  no nesting.


### Now

In Nextclade v3 we are trying to be more (or, hopefully, mostly) spec-compliant and we use `ID` and `Name` as the primary attributes. `ID` is for identification and `Name` for friendly display. The `ID` and `Parent` attributes are also used to deduce a nested hierarchy of features in order to support multiple CDS per gene, multiple CDS segments per CDS (for things like splicing and slippage) and multiple proteins per CDS (proteins are currently parsed but not yet used). The translation unit is a CDS and this is the most important feature type in Nextclade. Currently supported feature types are: gene, CDS, protein.

For backwards compatibility, reducing breakage when transitioning to v3, and in order to support various non-compliant GFF3 files in the wild (people are not always disciplined and upload to databases files that are not compliant) I allowed myself to keep reading `gene` and `gene_name` attributes as aliases for `Name`. Which means some datasets out there might have non-compliant genome annotations and work just fine.

I believe that as a part of this PR, changing  `gene` to `gene_name` or vice-versa has no effect. As they are both just aliases for `Name` ([src](https://github.com/nextstrain/nextclade/blob/3475c1be3b32015f405cbb9f5ecfe469b8754fce/packages/nextclade/src/io/gff3.rs#L14-L73)). But let me know if it's not working as intended.


Another important backwards-compat and non-compliant behavior Nextclade has:
 - if there's no gene for a CDS, it generates a gene with the same parameters
 - if there's no CDS for a gene, it generates a CDS with the same parameters

This is important in practice, because there are many pathogens where CDS-to-gene relationship is 1:1 and there is only genes or only CDS in GFF3 and people don't bother duplicating the info  (one line for every gene and one line for every CDS, which are the same as genes but with a different feature type), just to produce a compliant annotation.
 


### Debugging

1. You can use `read-annotation` CLI command to view a GFF3 file from the perspective of Nextclade. This will display a table with the hierarchy of features as understood by Nextclade. Only gene, CDS and protein features will be displayed, because these are the only ones Nextclade understands (proteins are not used currently).

	```bash
	nextclade read-annotation genome_annotation.gff3
	```

2. If you add `--json` flag, it will print the same information in JSON format:
 
	```bash
	nextclade read-annotation --json genome_annotation.gff3 > genome_annotation.json
	```
	
	The format is a dump of internal `struct`s and comes without any kind of stability guarantees. It is not recommended for use anywhere outside of debugging and can change any time. This format is used internally to figure out everything related to translation and amino acids. This is what you see in Nextclade Web on results page at the bottom of the table and in the gene selection dropdown in the header of the last column of the table.

3. You can also visualize ALL features from a GFF3 file by adding `--feature-tree` flag (not very intuitively named).

	```bash
	nextclade read-annotation genome_annotation.gff3 --feature-tree
	```
	
	All features are displayed, even the ones Nextclade does not understand. This is a precursor of the format we saw just above - it comes earlier in the parsing process, before unused/unknown feature types are filtered out and before any preprocessing and additional calculations. This aims to show the annotation as is, but in more readable way, so this command might be useful for visualizing GFF3 files more generally, even outside of context of Nextclade.

4. The `--json` flag also applies with `--feature-tree`. This is a different internal format which comes earlier in the parsing process. The format is also an unstable internal format, purely for debugging purposes.

**Comment by ivan-aksamentov** [comment]
> [The official Nextclade gene map docs](https://docs.nextstrain.org/projects/nextclade/en/stable/user/input-files/03-genome-annotation.html) suggest we should use CDS features with gene ids.

I believe it's a bug in the docs. The key should be `Name`, not `gene`

**Comment by ivan-aksamentov** [comment]
@genehack After seeing the error message in [`4eeffd8` (#3)](https://github.com/nextstrain/seasonal-cov/pull/3/commits/4eeffd8d9ef0b2d0205c33a12a32ddf82e5d20a9), I was looking into what could be wrong with Nextclade.

I assembled a dataset from the old ref and annotation files from here https://github.com/nextstrain/seasonal-cov/tree/8548b683c3afb953ebc067087d8fe80e553e7c96/phylogenetic/config/hku1 (previous commit if you click the "History" button in the GFF3 file on GitHub).
Here is the result. It's a .zip, but github does not allow uploading zips, so remove the .txt extension and unzip: [cov-hku-1.zip.txt](https://github.com/nextstrain/seasonal-cov/files/15342689/cov-hku-1.zip.txt)

Tried to visualize the annotation:

```bash
nextclade read-annotation cov-hku-1/genome_annotation.gff3
```


Tried to run with individual inputs (input fasta is just the reference fasta, because I cannot find any other example sequences):

```bash
nextclade run --input-ref cov-hku-1/reference.fasta --input-annotation cov-hku-1/genome_annotation.gff3 -O tmp/cov-hku-1 cov-hku-1/reference.fast
```

and tried to run as a dataset (it is possible because I added a minimal pathogen.json file there):

```bash
nextclade run --input-dataset cov-hku-1/ -O tmp/cov-hku-1 cov-hku-1/reference.fasta 
```

Everything seems to be working smoothly, so I cannot reproduce the error. Could there be a problem with Nextclade CLI version? Perhaps older version had some bugs.  I am not sure how Nextclade CLI is installed in this repo and what the version is. I used Nextclade CLI 3.5.0.

What the error in your commit message means is Nextclade perceives the entry for feature "N" (gene/CDS "N") as duplicated and it does not know which one to pick. Though in the GFF3 file I don't see this happening - there seems to be only one entry for "N". So this is very puzzling.


<details>
<summary>
Full working-as-intended repro in docker, using the cov-hku-1.zip.txt file from above (click to expand)
</summary>

```bash
docker run -it --rm nextstrain/nextclade:3.5.0 bash -c "set -euxo pipefail \
&& apt-get update -yqq && apt-get -yqq install curl unzip \
&& curl -fsSL -o cov-hku-1.zip https://github.com/nextstrain/seasonal-cov/files/15342689/cov-hku-1.zip.txt \
&& mkdir -p cov-hku-1; cd cov-hku-1; unzip ../cov-hku-1.zip; cd .. \
&& nextclade read-annotation cov-hku-1/genome_annotation.gff3 \
&& nextclade run --verbose --input-ref cov-hku-1/reference.fasta --input-annotation cov-hku-1/genome_annotation.gff3 -O tmp/cov-hku-1/individual cov-hku-1/reference.fasta \
&& ls -al tmp/cov-hku-1/individual/ \
&& cat tmp/cov-hku-1/individual/nextclade.tsv \
&& cat tmp/cov-hku-1/individual/nextclade.cds_translation.S1.fasta \
&& nextclade run --verbose --input-dataset cov-hku-1/ -O tmp/cov-hku-1/dataset cov-hku-1/reference.fasta \
&& ls -al tmp/cov-hku-1/dataset/ \
&& cat tmp/cov-hku-1/dataset/nextclade.tsv \
&& cat tmp/cov-hku-1/dataset/nextclade.cds_translation.S1.fasta \
"
```

</details>

**Comment by genehack** [comment]
@ivan-aksamentov thanks for looking into this. 

I checked out that same SHA ([8548b683c](https://github.com/nextstrain/seasonal-cov/tree/8548b683c3afb953ebc067087d8fe80e553e7c96/phylogenetic/config/hku1)) and ran the `ingest` and `phylogenetic` workflows, and I no longer see the error either. 

My earlier errors happened with an ambient runtime, which I ended up abandoning due to some weird dependency issues (probably triggered by installing some additional tooling on top of the initial Nextstrain build). With the Nextstrain Conda runtime I'm using now, I was able to run both workflows without issue.

**Pull request merged**